plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java

```diff
@@ -112,56 +117,75 @@ private static List<Page> buildPages(ConnectorTableMetadata tableMetadata, Table
-                pagesBuilder.appendIntegerBigintMap(dataFile.nanValueCounts());
-            }
-            if (checkNonNull(dataFile.lowerBounds(), pagesBuilder)) {
-                pagesBuilder.appendIntegerVarcharMap(dataFile.lowerBounds().entrySet().stream()
-                        .filter(entry -> idToTypeMapping.containsKey(entry.getKey()))
-                        .collect(toImmutableMap(
-                                Map.Entry<Integer, ByteBuffer>::getKey,
-                                entry -> Transforms.identity().toHumanString(
-                                        idToTypeMapping.get(entry.getKey()), Conversions.fromByteBuffer(idToTypeMapping.get(entry.getKey()), entry.getValue())))));
-            }
-            if (checkNonNull(dataFile.upperBounds(), pagesBuilder)) {
-                pagesBuilder.appendIntegerVarcharMap(dataFile.upperBounds().entrySet().stream()
-                        .filter(entry -> idToTypeMapping.containsKey(entry.getKey()))
-                        .collect(toImmutableMap(
-                                Map.Entry<Integer, ByteBuffer>::getKey,
-                                entry -> Transforms.identity().toHumanString(
-                                        idToTypeMapping.get(entry.getKey()), Conversions.fromByteBuffer(idToTypeMapping.get(entry.getKey()), entry.getValue())))));
-            }
-            if (checkNonNull(dataFile.keyMetadata(), pagesBuilder)) {
-                pagesBuilder.appendVarbinary(Slices.wrappedBuffer(dataFile.keyMetadata()));
-            }
-            if (checkNonNull(dataFile.splitOffsets(), pagesBuilder)) {
-                pagesBuilder.appendBigintArray(dataFile.splitOffsets());
-            }
-            if (checkNonNull(dataFile.equalityFieldIds(), pagesBuilder)) {
-                pagesBuilder.appendIntegerArray(dataFile.equalityFieldIds());
-            }
-            pagesBuilder.endRow();
-        });
+        Iterable<FileScanTask> planFiles = tableScan.planFiles();
+        Iterator<FileScanTask> fileScanTaskIterator = planFiles.iterator();
+
+        // Estimate the approximate memory size required
+        for (int i = 0; i < SCAN_FILE_BATCH_SIZE && fileScanTaskIterator.hasNext(); i++) {
+            addRow(pagesBuilder, fileScanTaskIterator.next(), idToTypeMapping);
+        }
 
-        return pagesBuilder.build();
+        if (!fileScanTaskIterator.hasNext()) {
+            return new FixedPageSource(pagesBuilder.build());
+        }
+
+        long estimatedMemoryUsageBytes = 0;
+        for (Page page : pagesBuilder.build()) {
+            estimatedMemoryUsageBytes += page.getRetainedSizeInBytes();
+        }
+
+        Iterable<Page> pages = new FilesTableIterable(planFiles, idToTypeMapping, tableMetadata);
+        return new FixedPageSource(pages, Optional.of(estimatedMemoryUsageBytes));
+    }
+
+    private static void addRow(PageListBuilder pagesBuilder, FileScanTask fileScanTask, Map<Integer, Type> idToTypeMapping)
```

Comment:
Now that `addRow` is called at the bottom of the file, put this at the bottom below where it's called.

Reply:
You'll need to change it a little to work with the `PageBuilder` API instead of `PageListBuilder` too